### ISSUE TITLE
Great project!

### ISSUE THREAD

**Issue by dibyendumajumdar** [freewilll/wcc#3]
Hi, I came across your project while researching / implementing some of the compiler algorithms in Engineering a Compiler.

Cool project! I love that you have written unit tests for various components, using examples from the book. I have been doing the same!

**Comment by freewilll** [comment]
Thanks for your feedback :). The project started off as a PoC so that I could learn about SSA and code generation. For example, all of the SSA transformations are there, but I never got around to any optimizations such as partial redundancy elimination. The book was fantastic and completing this compiler has been rewarding beyond belief.

**Comment by dibyendumajumdar** [comment]
I am learning how to do compiler optimizations and program analysis - and curiously I too started with the SSA translation without implementing any optimizer passes, because I wanted to first get the infrastructure there. I am currently working on a vanilla Chaitin register allocator.

**Comment by freewilll** [comment]
Doing the register allocation was so much fun. I had read of graph coloring and it all seemed difficult. But once I implemented it, it turned out to be pretty straight forward. I underestimated the code generation part big time. There were some pitfalls I didn't see coming and my early versions often produced incorrect code.

**Comment by dibyendumajumdar** [comment]
That is good to know! 

I will look at the Briggs version of the reg allocator at some point. 
Almost everyone appears to be using this rather than the original Chaitin algo.

**Comment by freewilll** [comment]
Interesting, [Briggs's thesis](https://www.cs.utexas.edu/~mckinley/380C/lecs/briggs-thesis-1992.pdf) mentions the committee included Cooper & Torczon, who wrote [Engineering a Compiler 2nd Edition](https://www.amazon.com/Engineering-Compiler-Keith-Cooper/dp/012088478X).

**Comment by dibyendumajumdar** [comment]
Indeed, Briggs did his PhD under Cooper & Torczon, and the algo for graph coloring described in the book is really from Brigg's thesis. 

I also implemented SSA based on Brigg's paper [Practical Improvements to SSA](https://homes.luddy.indiana.edu/achauhan/Teaching/B629/2006-Fall/CourseMaterial/1998-spe-briggs-ssa_improv.pdf). Again this is also the basis for the description in the book.

**Comment by dibyendumajumdar** [comment]
The other major pioneer Cliff Click who created the Sea of Nodes IR (an SSA based graph IR) - also did his Phd under Cooper & Torczon.

https://github.com/seaofnodes
